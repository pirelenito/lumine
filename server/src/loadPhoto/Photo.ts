export default interface Photo {
  relativePath: string
  contentHash: string

  // preview: {
  //   thumbnailPath: string
  //   fullsizePath: string
  // }

  // metadata: {
  //   file: {
  //     ctimeMs: number
  //     mtimeMs: number
  //     size: number
  //   }
  // }
}
