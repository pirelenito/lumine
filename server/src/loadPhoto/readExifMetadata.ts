import { pickBy, type, contains } from 'ramda'
import { ExifImage } from 'exif'
import { ExifMetadata } from './Photo'

const onlyNumbersAndStrings = pickBy((val, key) => contains(type(val), ['Number', 'String']))

export default (fullPath: string): Promise<ExifMetadata> =>
  new Promise((resolve, reject) => {
    try {
      new ExifImage({ image: fullPath }, (err: any, metadata = { image: {}, exif: {}, gps: {} }) => {
        if (err && err.message !== 'No Exif segment found in the given image.') {
          return reject(err)
        }

        const rawMetadata = {
          image: onlyNumbersAndStrings(metadata.image),
          exif: onlyNumbersAndStrings(metadata.exif),
          gps: onlyNumbersAndStrings(metadata.gps),
        }

        resolve(rawMetadata)
      })
    } catch (e) {
      reject(e)
    }
  })
