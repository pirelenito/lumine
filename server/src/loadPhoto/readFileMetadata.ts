import { FileMetadata } from './Photo'

const promisify = require('util').promisify
const stat = promisify(require('fs').stat)

export default async (fullPath: string): Promise<FileMetadata> => {
  const fileStat = await stat(fullPath)

  return {
    // birthtime can be null on some filesystems
    // which makes node return something else that might be wrong
    createdAt: convertToDate(fileStat.birthtimeMs > fileStat.mtimeMs ? fileStat.mtimeMs : fileStat.birthtimeMs),
    modifiedAt: convertToDate(fileStat.mtimeMs),
    size: fileStat.size,
  }
}

const convertToDate = (timeMs: number) => new Date(Math.floor(timeMs / 1000) * 1000)
