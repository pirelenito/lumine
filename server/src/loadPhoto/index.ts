import Photo from './Photo'
import { join } from 'path'
import generatePreview from './generatePreview'
import calculateContentHash from './calculateContentHash'
import readFileMetadata from './readFileMetadata'

interface Config {
  libraryBasePath: string
  cacheBasePath: string
}

export default (config: Config) => async (relativePath: string): Promise<Photo> => {
  const fullPath = join(config.libraryBasePath, relativePath)
  const contentHash = await calculateContentHash(fullPath)
  const fileMetadata = await readFileMetadata(fullPath)
  const preview = await generatePreview(config.cacheBasePath)(contentHash, fullPath)

  return { relativePath, contentHash, metadata: { file: fileMetadata }, preview }
}
