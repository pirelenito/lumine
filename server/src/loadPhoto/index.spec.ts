import loadPhoto from '.'
import { join } from 'path'

it('loads photo information', async () => {
  const config = { libraryBasePath: join(__dirname, 'fixtures'), cacheBasePath: join(__dirname, 'cache') }

  const photo = await loadPhoto(config)('DSC00010.ARW')

  expect(photo).toEqual({
    contentHash: '98ce370e686f5a40a02523058679a06ca3ff4ce2',
    metadata: { file: { createdAt: 1542831393, modifiedAt: 1534248554, size: 25133056 } },
    preview: {
      fullSizePath: 'fullSize/98/ce370e686f5a40a02523058679a06ca3ff4ce2.jpg',
      thumbnailPath: 'thumbnail/98/ce370e686f5a40a02523058679a06ca3ff4ce2.jpg',
    },
    relativePath: 'DSC00010.ARW',
  })
}, 180000)
