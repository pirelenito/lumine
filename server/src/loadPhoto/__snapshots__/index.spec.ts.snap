// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads a6000 1`] = `
Object {
  "contentHash": "98ce370e686f5a40a02523058679a06ca3ff4ce2",
  "metadata": Object {
    "exif": Object {
      "exif": Object {
        "BrightnessValue": 8.946875,
        "ColorSpace": 1,
        "CompressedBitsPerPixel": 8,
        "Contrast": 0,
        "CreateDate": "2018:08:14 14:09:14",
        "CustomRendered": 0,
        "DateTimeOriginal": "2018:08:14 14:09:14",
        "DigitalZoomRatio": 1,
        "ExifImageHeight": 4000,
        "ExifImageWidth": 6000,
        "ExposureCompensation": 0,
        "ExposureMode": 0,
        "ExposureProgram": 3,
        "ExposureTime": 0.00125,
        "FNumber": 3.2,
        "Flash": 16,
        "FocalLength": 35,
        "FocalLengthIn35mmFormat": 52,
        "ISO": 100,
        "InteropOffset": 38114,
        "LensModel": "E 35mm F1.8 OSS",
        "LightSource": 0,
        "MaxApertureValue": 1.6953125,
        "MeteringMode": 5,
        "RecommendedExposureIndex": 100,
        "Saturation": 0,
        "SceneCaptureType": 0,
        "SensitivityType": 2,
        "Sharpness": 0,
        "WhiteBalance": 0,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 348,
        "ImageDescription": "                               ",
        "Make": "SONY",
        "Model": "ILCE-6000",
        "ModifyDate": "2018:08:14 14:09:14",
        "Orientation": 1,
        "ProcessingSoftware": "UFRaw 0.20",
        "Software": "ILCE-6000 v3.20",
        "SubfileType": 1,
        "YCbCrPositioning": 2,
      },
    },
    "file": Object {
      "createdAt": 1542836886,
      "modifiedAt": 1534248554,
      "size": 25133056,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/98/ce370e686f5a40a02523058679a06ca3ff4ce2.jpg",
    "thumbnailPath": "thumbnail/98/ce370e686f5a40a02523058679a06ca3ff4ce2.jpg",
  },
  "relativePath": "a6000.ARW",
}
`;

exports[`loads iphone 6 1`] = `
Object {
  "contentHash": "e56e56185e42ac9569b4d811adb5cd495fbcebec",
  "metadata": Object {
    "exif": Object {
      "exif": Object {
        "ApertureValue": 2.2750071245369052,
        "BrightnessValue": 11.340301003344482,
        "ColorSpace": 1,
        "CreateDate": "2018:08:09 18:06:30",
        "DateTimeOriginal": "2018:08:09 18:06:30",
        "ExifImageHeight": 1668,
        "ExifImageWidth": 2968,
        "ExposureCompensation": 0,
        "ExposureMode": 0,
        "ExposureProgram": 2,
        "ExposureTime": 0.00032797638570022957,
        "FNumber": 2.2,
        "Flash": 24,
        "FocalLength": 4.15,
        "FocalLengthIn35mmFormat": 36,
        "ISO": 32,
        "LensMake": "Apple",
        "LensModel": "iPhone 6 back camera 4.15mm f/2.2",
        "MeteringMode": 5,
        "SceneCaptureType": 0,
        "SensingMethod": 2,
        "ShutterSpeedValue": 11.574016563146998,
        "SubSecTimeDigitized": "093",
        "SubSecTimeOriginal": "093",
        "WhiteBalance": 0,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 194,
        "Make": "Apple",
        "Model": "iPhone 6",
        "ModifyDate": "2018:08:09 18:06:30",
        "Orientation": 1,
        "ResolutionUnit": 2,
        "Software": "11.4.1",
        "XResolution": 72,
        "YCbCrPositioning": 1,
        "YResolution": 72,
      },
    },
    "file": Object {
      "createdAt": 1542836883,
      "modifiedAt": 1533830790,
      "size": 993124,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/e5/6e56185e42ac9569b4d811adb5cd495fbcebec.jpg",
    "thumbnailPath": "thumbnail/e5/6e56185e42ac9569b4d811adb5cd495fbcebec.jpg",
  },
  "relativePath": "iphone6.jpg",
}
`;

exports[`loads nokia 5 1`] = `
Object {
  "contentHash": "da1fb28e2764fcd4edc27c4c65167a1349c9b802",
  "metadata": Object {
    "exif": Object {
      "exif": Object {
        "ApertureValue": 2,
        "BrightnessValue": 0.3,
        "ColorSpace": 1,
        "CreateDate": "2018:11:08 12:41:58",
        "DateTimeOriginal": "2018:11:08 12:41:58",
        "ExifImageHeight": 3120,
        "ExifImageWidth": 4160,
        "ExposureMode": 0,
        "ExposureProgram": 0,
        "ExposureTime": 0.030303030303030304,
        "FNumber": 2,
        "Flash": 16,
        "FocalLength": 3.65,
        "FocalLengthIn35mmFormat": 0,
        "ISO": 320,
        "InteropOffset": 759,
        "MeteringMode": 1,
        "SceneCaptureType": 0,
        "SensingMethod": 2,
        "ShutterSpeedValue": 5.059,
        "SubSecTime": "931367",
        "SubSecTimeDigitized": "931367",
        "SubSecTimeOriginal": "931367",
        "WhiteBalance": 0,
      },
      "gps": Object {
        "GPSAltitudeRef": 2,
        "GPSDateStamp": "2018:11:08",
      },
      "image": Object {
        "ExifOffset": 242,
        "GPSInfo": 789,
        "Make": "",
        "Model": "Nokia 5",
        "ModifyDate": "2018:11:08 12:41:58",
        "ResolutionUnit": 2,
        "Software": "TA-1053_00WW-user 8.1.0 OPR1.170623.026 00WW_5_63D release-keys",
        "XResolution": 72,
        "YCbCrPositioning": 1,
        "YResolution": 72,
      },
    },
    "file": Object {
      "createdAt": 1542836846,
      "modifiedAt": 1542836831,
      "size": 2273365,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/da/1fb28e2764fcd4edc27c4c65167a1349c9b802.jpg",
    "thumbnailPath": "thumbnail/da/1fb28e2764fcd4edc27c4c65167a1349c9b802.jpg",
  },
  "relativePath": "nokia5.jpg",
}
`;
